rust/cyclotron-core/src/base_ops.rs

```diff
@@ -0,0 +1,748 @@
+SELECT *
+FROM UNNEST(
+        $1,
+        $2,
+        $3,
+        $4,
+        $5,
+        $6,
+        $7,
+        $8,
+        $9,
+        $10,
+        $11,
+        $12,
+        $13,
+        $14,
+        $15,
+        $16,
+        $17
+    )
+"#,
+    )
+    .bind(ids)
+    .bind(team_ids)
+    .bind(function_ids)
+    .bind(created_at)
+    .bind(lock_ids)
+    .bind(last_heartbeats)
+    .bind(janitor_touch_counts)
+    .bind(transition_counts)
+    .bind(last_transitions)
+    .bind(queue_names)
+    .bind(waiting_ons)
+    .bind(states)
+    .bind(scheduleds)
+    .bind(priorities)
+    .bind(vm_states)
+    .bind(metadatas)
+    .bind(parameters)
+    .execute(executor)
+    .await?;
+
+    Ok(())
+}
+
+// Dequeue the next job batch from the queue, skipping VM state since it can be large
+pub async fn dequeue_jobs<'c, E>(
+    executor: E,
+    queue: &str,
+    worker_type: WaitingOn,
+    max: usize,
+) -> Result<Vec<Job>, QueueError>
+where
+    E: sqlx::Executor<'c, Database = sqlx::Postgres>,
+{
+    // TODO - right now, locks are completely transient. We could instead have the lock_id act like a
+    // "worker_id", and be provided by the caller, which would let workers do less bookkeeping, and make
+    // some kinds of debugging easier, but I prefer locks being opaque to workers for now, to avoid any
+    // confusion or potential for accidental deadlocking (e.g. if someone persisted the worker_id across
+    // process restarts).
+    let lock_id = Uuid::now_v7();
+    Ok(sqlx::query_as!(
+        Job,
+        r#"
+WITH available AS (
+    SELECT
+        id,
+        state
+    FROM cyclotron_jobs
+    WHERE
+        state = 'available'::JobState
+        AND waiting_on = $1
+        AND queue_name = $2
+        AND scheduled <= NOW()
+    ORDER BY
+        priority ASC,
+        scheduled ASC
+    LIMIT $3
+    FOR UPDATE SKIP LOCKED
+)
+UPDATE cyclotron_jobs
+SET
+    state = 'running'::JobState,
+    lock_id = $4,
+    last_heartbeat = NOW(),
+    last_transition = NOW(),
+    transition_count = transition_count + 1
+FROM available
+WHERE
+    cyclotron_jobs.id = available.id
+RETURNING
+    cyclotron_jobs.id,
+    team_id,
+    available.state as "state: JobState",
+    waiting_on as "waiting_on: WaitingOn",
+    queue_name,
+    priority,
+    function_id,
+    created,
+    last_transition,
+    scheduled,
+    transition_count,
+    NULL as vm_state,
+    metadata,
+    parameters,
+    lock_id,
+    last_heartbeat,
+    janitor_touch_count
+    "#,
+        worker_type as _,
+        queue,
+        max as i64,
+        lock_id
+    )
+    .fetch_all(executor)
+    .await?)
+}
+
+// Dequeue a batch of jobs, also returning their VM state. This is an optimisation - you could
+// dequeue a batch of jobs and then fetch their VM state in a separate query, but this is hopefully less
+// heavy on the DB, if a given worker knows it needs VM state for all dequeue jobs
+pub async fn dequeue_with_vm_state<'c, E>(
```

Comment:
Yep, the sqlx macro requires the string to be embedded, rather than a ref, unfortunately. I wondered if I could do something compile-time like the `include_str`, but didn't look into it